cs/src/Management/TunnelManagementClient.cs

```diff
@@ -924,6 +925,60 @@ public void Dispose()
             CancellationToken cancellation)
         {
             Requires.NotNull(tunnel, nameof(tunnel));
+            options ??= new TunnelRequestOptions();
+            options.AdditionalHeaders ??= new List<KeyValuePair<string,string>>();
+            options.AdditionalHeaders = options.AdditionalHeaders.Append(new KeyValuePair<string, string>("If-None-Match", "*"));
+            var tunnelId = tunnel.TunnelId;
+            var idGenerated = string.IsNullOrEmpty(tunnelId);
+            if (idGenerated)
+            {
+                tunnel.TunnelId = IdGeneration.GenerateTunnelId();
+            }
+            for (int retries = 0; retries <= 3; retries++)
+            {
+                try
+                {
+                    var result = await this.SendTunnelRequestAsync<Tunnel, Tunnel>(
+                       HttpMethod.Put,
+                       tunnel,
+                       ManageAccessTokenScope,
+                       path: null,
+                       query: GetApiQuery(),
+                       options,
+                       ConvertTunnelForRequest(tunnel),
+                       cancellation,
+                       true);
+                    PreserveAccessTokens(tunnel, result);
+                    return result!;
+                }
+                catch (UnauthorizedAccessException) when (idGenerated && retries < 3) // The tunnel ID was already taken.
```

Comment:
Could be worth putting 3 into a constant. Or, change `retries` to `remainingRetries`, initialize to 3, then decrement.

Reply:
update